### DISCUSSION TITLE
(New in v4.5.0) Pivot Table

### DISCUSSION THREAD

**Discussion by mikez** [exceljs/exceljs#2575] [General]
Pivot Table support is coming with ExcelJS 4.5.0.
(As of ExcelJS 4.4.1, experimental support is available.)

## Try it out

```js
const workbook = new Excel.Workbook();

const worksheet1 = workbook.addWorksheet('Sheet1');
worksheet1.addRows([
  ['A', 'B', 'C', 'D', 'E'],
  ['a1', 'b1', 'c1', 4, 5],
  ['a1', 'b2', 'c1', 4, 5],
  ['a2', 'b1', 'c2', 14, 24],
  ['a2', 'b2', 'c2', 24, 35],
  ['a3', 'b1', 'c3', 34, 45],
  ['a3', 'b2', 'c3', 44, 45],
]);

const worksheet2 = workbook.addWorksheet('Sheet2');
worksheet2.addPivotTable({
  // Source data: entire sheet range; header in first row
  sourceSheet: worksheet1,
  // Pivot table fields: via header row in `worksheet1`
  rows: ['A', 'B'],
  columns: ['C'],
  values: ['E'], // Exactly 1 field
  metric: 'sum', // Metric: 'sum' only
});
```

Note that the names `rows`, `columns`, and `values` are modeled after the Pivot Table Fields dialogue in Excel:

![image](https://github.com/exceljs/exceljs/assets/216956/52defe6c-21c3-4597-bffd-c33189712405)

## Limitations

The current experimental version has several limitations, including:

- **No read support**. You can't read xlsx documents with existing pivot tables. Writing is supported.
- **One value only with "sum"**. Put as many rows and columns as you want, but you can only use one "value" item with the sum metric.
- **One pivot table per document**. Only one pivot table can be added for the entire document.

## Feedback

Please try it out and leave feedback of how it goes. Also feel free to request features.

**Comment by Rablet** [comment]
Excellent timing as I needed this for a project I'm working on.
I submitted a PR for supporting multiple values (as long as custom columns are not used): https://github.com/exceljs/exceljs/pull/2578

**Reply by mikez** [reply]
Your PR handles cases like these, right?
```js
{…, columns: [], values: ['D', 'E'], metric: 'sum'}
```

**Reply by Rablet** [reply]
Exactly! I should have added a sample to make that clearer. 

For me I needed this as I have a Pivot Table where I have two different prices that I want to sum up and I am not interested in any additional columns. It works well for that.
I would have added multi value + multi column support but wasn't sure how to debug when that scenario kept breaking.

**Comment by vaynevayne** [comment]
Will 4.4.1 also include experimental support for nest Columns?

**Reply by mikez** [reply]
Can you add a video of what you mean by "nest Columns"?
(If you do a `git clone https://github.com/exceljs/exceljs` you can try it out. See e.g. [this file](https://github.com/exceljs/exceljs/blob/master/test/test-pivot-table.js).)

**Reply by vaynevayne** [reply]
https://github.com/exceljs/exceljs/pull/1889

**Reply by mikez** [reply]
If I understand [your linked commit](https://github.com/exceljs/exceljs/pull/1889) correctly, it merges cells together to create the "appearance" of nested columns.
In [Pivot tables](https://support.microsoft.com/en-au/office/create-a-pivottable-to-analyze-worksheet-data-a9a84538-bfe9-40a9-a8e9-f99134456576#ID0EBL=Mac) (another concept), the columns are by design nested. Try it out! :)

**Reply by vaynevayne** [reply]
When will 4.4.1 be pushed? So I can try it

**Reply by mikez** [reply]
You can try it today if you do `git clone https://github.com/exceljs/exceljs`.

**Reply by vaynevayne** [reply]
There's no playground, and yarn add https://xxx doesn't work. It's tough
but pivot table  api should link https://ant.design/components/table#components-table-demo-grouping-columns

`columns[n].children`,

**Reply by mikez** [reply]
```js
node test/test-pivot-table.js yourfilename.xlsx
```

Have you understood the concept of pivot tables? It's used for data analysis. Grouping columns is another concept.

**Comment by shubhamgoswamiofficial** [comment]
From where shall i get version 4.4.1?  As on npm latest is 4.4.0.
**Bit urgent**

**Reply by hung-payerprice** [reply]
+1 I'm also interested in this pivot table functionality. Is there any update on the timeline or ETA of 4.4.1 npm package release? Thank you

**Reply by mikez** [reply]
FYI @shubhamgoswamiofficial @hung-payerprice @Rablet
I'm not the maintainer of this library. I had implemented Pivot Tables elsewhere and contributed a first version of it to this library for the benefit of us all. The current maintainer (@Siemienik) asked me to post here so you could try it out and leave feedback. To reach the maintainer or contribute to a release, you may try Discord: https://discord.gg/siema

**Reply by Siemienik** [reply]
[Discord](https://discord.gg/siema) or my email consulting@siemienik.com address is good place to catch me with urgent tickets.

I have got plans to release 4.4.1 with current progress soon (in march if everything would match the plan) 

Please keep in mind, that's still experimental feature with limitations.

**Reply by RaphyPoo** [reply]
Hey, I am trying to implement the pivot tables into one of my projects but I see that it has not yet been released. Is there an ETA to when this will come out? I need it for a project that I am currently working on. I really appreciate the work you guys do!

**Reply by gogromat** [reply]
you can install `exceljs` at specific commit: (npm|yarn|etc.): `yarn add https://github.com/exceljs/exceljs#ddab279`
add a "postinstall" to `package.json` "scripts" section: `"cd node_modules/exceljs/ && yarn install && yarn run build"`

**Reply by oitozero** [reply]
@Siemienik will 4.4.1 be released?

**Comment by OlegMuzych** [comment]
Hello there! What is the difficulty in making pivot table reading support? Why when reading a file .xlsx and its saving are the pivot tables missing? In no library has npm found support for reading and saving linked tables.

**Comment by gogromat** [comment]
Does not take columns or values when name contains newline characters (`\r\n`).
Ex. `"Time\r\nWorked"`.
Removing newline characters produces correct pivot table.

**Reply by mikez** [reply]
@gogromat Thanks for reporting! Do you see an easy way to fix this in the code? Alas, the maintainer of this repository seems to have gone missing. But maybe he will return eventually.

**Comment by dsilva01** [comment]
Hey everyone,

I’ve submitted a PR to add support for the Count metric in PivotTable, which allows users to aggregate data based on the number of occurrences in a given field. Currently, exceljs supports **Sum** metric, but **Count** was missing. This update brings it closer to native Excel functionality. https://github.com/exceljs/exceljs/pull/2885

**Comment by alirezaalavi87** [comment]
Hi.
When trying to open the exported XLSX file, excel shows an error about how the file is malformed. if you choose to recover it, the excel opens and then you have to press "enable editing", excel shows this message:
```
Excel completed file level validation and repair. Some parts of this workbook may have been repaired or discarded.
Repaired Records: PivotTable report from /xl/pivotTables/pivotTable1.xml part (PivotTable view)
```
Then the pivot table gets displayed.

What's the issue how can we fix it?

**Reply by dsilva01** [reply]
@alirezaalavi87  can you show the code?

**Reply by alirezaalavi87** [reply]
of course!
I am using it in [Nu-pivottables](https://github.com/alirezaalavi87/nu-pivottables) `spreadsheet_renderers.js`.

https://github.com/alirezaalavi87/nu-pivottables/blob/03d80ec7775f0fe12629557697874c8689a65c9d/src/renderers/spreadsheet_renderers.js

**Reply by protobi-pieter** [reply]
@alirezaalavi87 Try https://github.com/protobi/exceljs as this has some recent extensions to Pivot table generation handling special cases such as blank/missing values, escaping ampersands in text strings, etc.
